src/Core/Managed/Shared/TelemetryClient.cs

```diff
@@ -17,14 +17,19 @@
     /// <summary>
     /// Send events, metrics and other telemetry to the Application Insights service.
     /// </summary>
-    public sealed class TelemetryClient
+    public sealed class TelemetryClient : IDisposable
```

Comment:
I don't think `TelemetryClient` should be managing the `TelemetryConfiguration.TelemetryChannel` or the `TelemetryConfiguration`. It doesn't own these objects; they are given to the `TelemetryClient` constructor by their owner. Disposing `TelemetryChannel` should be done by its proper owner; in case of #364 I believe its the user code. 

#364 is caused by a user error. The proper behavior for `TelemetryChannel(TelemetryConfiguration)` would be to throw an `ArgumentException` when `TelemetryConfiguration.TelemetryChannel` is null. This should never be the case by default because `TelemetryConfiguration.Active.TelemetryChannel` is automatically initialized by the `TelemetryConfigurationFactory`. 

This may be reopening the old debate regarding whether an instrumentation API should throw exception when user messes up or try to handle these mistakes defensively in the SDK. I think this bug, the amount of effort required to keep the code running in face of user errors and the resulting API impact are a strong argument in favor of exceptions.